subprojects/plugins/src/main/java/org/gradle/api/plugins/WarPlugin.java

```diff
@@ -65,7 +68,8 @@ public void apply(final Project project) {
         project.getConvention().getPlugins().put("war", pluginConvention);
 
         project.getTasks().withType(War.class).configureEach(task -> {
-            task.from((Callable) () -> pluginConvention.getWebAppDir());
+            task.getWebAppDir().set(project.provider(() -> fileFactory.dir(pluginConvention.getWebAppDir())));
```

Comment:
This should use `convention` and I would either add an internal property to `WarPluginConvention` that's already a `DirectoryProperty` or use `ProjectLayout.dir` to convert from a `File` to a `Provider<Directory>`.